File: src/db-toolkit/operations/connection_manager.py
"""Connection management operations."""

import asyncio
from typing import Dict, List, Optional

from connectors.base import BaseConnector
from connectors.factory import ConnectorFactory
from core.models import DatabaseConnection
from core.settings_storage import SettingsStorage
from operations.operation_lock import operation_lock


class ConnectionManager:
    """Manages active database connections."""

    def __init__(self):
        """Initialize connection manager."""
        self._active_connections: Dict[str, BaseConnector] = {}
        self._connection_metadata: Dict[str, DatabaseConnection] = {}
        self._settings_storage = SettingsStorage()

    async def connect(self, connection: DatabaseConnection, timeout: Optional[int] = None) -> bool:
        """Establish database connection with timeout."""
        try:
            # Get timeout from settings if not provided
            if timeout is None:
                settings = await self._settings_storage.get_settings()
                timeout = settings.connection_timeout
            
            connector = ConnectorFactory.create_connector(connection.db_type)
            
            # Connect with timeout
            success = await asyncio.wait_for(
                connector.connect(connection),
                timeout=timeout
            )

            if success:
                self._active_connections[connection.id] = connector
                self._connection_metadata[connection.id] = connection
                return True
            return False

        except asyncio.TimeoutError:
            return False
        except Exception:
            return False

    async def disconnect(self, connection_id: str) -> bool:
        """Disconnect from database."""

        connector = self._active_connections.get(connection_id)
        if connector:
            success = await connector.disconnect()
            if success:
                del self._active_connections[connection_id]
                if connection_id in self._connection_metadata:
                    del self._connection_metadata[connection_id]
                operation_lock.cleanup(connection_id)
            return success
        return False

    async def get_connector(self, connection_id: str, auto_reconnect: bool = True) -> Optional[BaseConnector]:
        """Get active connector by connection ID with auto-reconnect."""
        connector = self._active_connections.get(connection_id)
        
        # Check if auto-reconnect is enabled and connection is lost
        if connector and not connector.is_connected and auto_reconnect:
            settings = await self._settings_storage.get_settings()
            if settings.auto_reconnect:
                connection = self._connection_metadata.get(connection_id)
                if connection:
                    # Try to reconnect
                    success = await self.connect(connection)
                    if success:
                        return self._active_connections.get(connection_id)
        
        return connector

    async def get_connection(self, connection_id: str) -> Optional[DatabaseConnection]:
        """Get connection metadata by ID."""
        return self._connection_metadata.get(connection_id)

    async def is_connected(self, connection_id: str) -> bool:
        """Check if connection is active."""
        connector = self._active_connections.get(connection_id)
        return connector is not None and connector.is_connected

    async def get_all_active_connections(self) -> List[str]:
        """Get list of all active connection IDs."""
        return list(self._active_connections.keys())

    async def get_connection_count(self) -> int:
        """Get count of active connections."""
        return len(self._active_connections)

    async def disconnect_all(self):
        """Disconnect all active connections."""
        for connection_id in list(self._active_connections.keys()):
            await self.disconnect(connection_id)

    async def get_connection_status(self, connection_id: str) -> Dict:
        """Get detailed connection status."""

        connector = self._active_connections.get(connection_id)
        metadata = self._connection_metadata.get(connection_id)

        if not connector or not metadata:
            return {"connected": False, "locked": False}

        return {
            "connected": connector.is_connected,
            "locked": operation_lock.is_locked(connection_id),
            "db_type": metadata.db_type.value,
            "name": metadata.name,
        }

connection_manager = ConnectionManager()

File: src/db-toolkit/electron/backend/utils/connection-manager.js
/**
 * Connection management operations.
 */

const { ConnectorFactory } = require('../connectors');
const { settingsStorage } = require('./settings-storage');
const { logger } = require('./logger');

class ConnectionManager {
  constructor() {
    this.activeConnections = new Map();
    this.connectionMetadata = new Map();
    this.reconnectTimers = new Map();
  }

  async connect(connection, timeout = null) {
    logger.info(`Connecting to '${connection.name}' (${connection.db_type})`);
    
    try {
      if (timeout === null) {
        const settings = await settingsStorage.getSettings();
        timeout = settings.connection_timeout;
      }
      
      const connector = ConnectorFactory.createConnector(connection.db_type);
      
      const connectPromise = connector.connect(connection);
      const timeoutPromise = new Promise((_, reject) =>
        setTimeout(() => reject(new Error('Connection timeout')), timeout * 1000)
      );
      
      const success = await Promise.race([connectPromise, timeoutPromise]);
      
      if (success) {
        this.activeConnections.set(connection.id, connector);
        this.connectionMetadata.set(connection.id, connection);
        logger.info(`Successfully connected to '${connection.name}'`);
        
        if (connection.auto_reconnect) {
          this.setupAutoReconnect(connection.id);
        }
        
        return true;
      }
      
      logger.error(`Failed to connect to '${connection.name}'`);
      return false;
    } catch (error) {
      logger.error(`Connection error for '${connection.name}': ${error.message}`);
      
      if (connection.auto_reconnect) {
        this.scheduleReconnect(connection.id, connection);
      }
      
      return false;
    }
  }

  async disconnect(connectionId) {
    logger.info(`Disconnecting from connection '${connectionId}'`);
    
    this.clearReconnectTimer(connectionId);
    
    const connector = this.activeConnections.get(connectionId);
    if (connector) {
      const success = await connector.disconnect();
      if (success) {
        this.activeConnections.delete(connectionId);
        this.connectionMetadata.delete(connectionId);
      }
      return success;
    }
    return false;
  }

  async getConnector(connectionId) {
    return this.activeConnections.get(connectionId);
  }

  async getConnection(connectionId) {
    return this.connectionMetadata.get(connectionId);
  }

  async isConnected(connectionId) {
    const connector = this.activeConnections.get(connectionId);
    return connector !== undefined && connector.isConnected;
  }

  async getAllActiveConnections() {
    return Array.from(this.activeConnections.keys());
  }

  async getConnectionCount() {
    return this.activeConnections.size;
  }

  async disconnectAll() {
    const connectionIds = Array.from(this.activeConnections.keys());
    for (const connectionId of connectionIds) {
      await this.disconnect(connectionId);
    }
  }

  setupAutoReconnect(connectionId) {
    const checkInterval = setInterval(async () => {
      const connector = this.activeConnections.get(connectionId);
      const connection = this.connectionMetadata.get(connectionId);
      
      if (!connector || !connection) {
        this.clearReconnectTimer(connectionId);
        return;
      }
      
      if (!connector.isConnected) {
        logger.warn(`Connection lost for '${connection.name}', attempting reconnect...`);
        await this.connect(connection);
      }
    }, 30000); // Check every 30 seconds
    
    this.reconnectTimers.set(connectionId, checkInterval);
  }

  scheduleReconnect(connectionId, connection, attempt = 1) {
    if (attempt > 3) {
      logger.error(`Max reconnect attempts reached for '${connection.name}'`);
      return;
    }
    
    const delay = Math.min(1000 * Math.pow(2, attempt - 1), 10000); // Exponential backoff, max 10s
    logger.info(`Scheduling reconnect attempt ${attempt} for '${connection.name}' in ${delay}ms`);
    
    const timer = setTimeout(async () => {
      const success = await this.connect(connection);
      if (!success) {
        this.scheduleReconnect(connectionId, connection, attempt + 1);
      }
    }, delay);
    
    this.reconnectTimers.set(connectionId, timer);
  }

  clearReconnectTimer(connectionId) {
    const timer = this.reconnectTimers.get(connectionId);
    if (timer) {
      clearTimeout(timer);
      clearInterval(timer);
      this.reconnectTimers.delete(connectionId);
    }
  }

  async getConnectionStatus(connectionId) {
    const connector = this.activeConnections.get(connectionId);
    const metadata = this.connectionMetadata.get(connectionId);

    if (!connector || !metadata) {
      return { connected: false };
    }

    return {
      connected: connector.isConnected,
      db_type: metadata.db_type,
      name: metadata.name,
    };
  }
}

const connectionManager = new ConnectionManager();

module.exports = { ConnectionManager, connectionManager };
